feat: add `math/base/special/powf`

Neerajpathak07 opened the pull request:
Progresses #649.

## Description

> What is the purpose of this pull request?

This pull request:

-   adds `math/base/special/powf` which is the single-precision version of `pow`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   progresses #649 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-  [x] Read, understood, and followed the [contributing guidelines](https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md)

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Neerajpathak07 committed:
feat: add basic files

Neerajpathak07 committed:
feat: add js implementation

Neerajpathak07 commented on lib/node_modules/@stdlib/math/base/special/powf/lib/main.js:
```diff
@@ -0,0 +1,384 @@
+
+// VARIABLES //
+
+// 0x3fefffff = 1072693247 => 0 01111111110 11111111111111111111 => biased exponent: 1022 = -1+1023 => 2^-1
+var HIGH_MAX_NEAR_UNITY = 0x3fefffff|0; // asm type annotation
+
+// 0x41e00000 = 1105199104 => 0 10000011110 00000000000000000000 => biased exponent: 1054 = 31+1023 => 2^31
+var HIGH_BIASED_EXP_31 = 0x41e00000|0; // asm type annotation
+
+// 0x43f00000 = 1139802112 => 0 10000111111 00000000000000000000 => biased exponent: 1087 = 64+1023 => 2^64
+var HIGH_BIASED_EXP_64 = 0x43f00000|0; // asm type annotation
+
+// 0x40900000 = 1083179008 => 0 10000001001 00000000000000000000 => biased exponent: 1033 = 10+1023 => 2^10 = 1024
+var HIGH_BIASED_EXP_10 = 0x40900000|0; // asm type annotation
+
+// 0x3ff00000 = 1072693248 => 0 01111111111 00000000000000000000 => biased exponent: 1023 = 0+1023 => 2^0 = 1
+var HIGH_BIASED_EXP_0 = 0x3ff00000|0; // asm type annotation
+
+// 0x4090cc00 = 1083231232 => 0 10000001001 00001100110000000000
+var HIGH_1075 = 0x4090cc00|0; // asm type annotation
```
these variables are yet to be scoped according to `float32` ranges will update it while working on the modules.

kgryte replied:
These variables are only necessary if the upstream implementation has them.

Neerajpathak07 commented on lib/node_modules/@stdlib/math/base/special/powf/lib/log2axf.js:
```diff
@@ -0,0 +1,212 @@
+var polyvalL = require( './polyval_l.js' );
+
+
+// VARIABLES //
```
these values have been based on upstream implementation of [`powf`](https://svnweb.freebsd.org/base/release/12.2.0/lib/msun/src/e_powf.c?revision=367086&view=markup) in msun.

Neerajpathak07 commented:
@kgryte @gunjjoshi what do you guys think of the JS implementation so far!!

Neerajpathak07 commented on lib/node_modules/@stdlib/math/base/special/powf/lib/index.js:
```diff
@@ -0,0 +1,61 @@
+* v = powf( 100.0, 0.0 );
+* // returns 1.0
+*
+* v = powf( 3.1415927410125732, 5.0 );
```
per return value of `@stdlib/constants/float32/pi`

Neerajpathak07 commented on lib/node_modules/@stdlib/math/base/special/powf/lib/main.js:
```diff
@@ -0,0 +1,384 @@
+var uint32ToInt32 = require( '@stdlib/number/uint32/base/to-int32' );
+var NINF = require( '@stdlib/constants/float32/ninf' );
+var PINF = require( '@stdlib/constants/float32/pinf' );
+var ABS_MASK = require( '@stdlib/constants/float32/high-word-abs-mask' );
```
@gunjjoshi like we discussed!
Modules like `@stdlib/constants/float32/high-word-abs-mask` are yet to have `float32` implementations along with `set-lo-word` and `toword`

kgryte replied:
@Neerajpathak07 You should spend some time reading up on the binary format of floating point numbers. to-words and other float64 operations don't make sense. Why? Because a single precision float fits in a single word (ie, a 32-bit integer).

gunjjoshi replied:
@Neerajpathak07 Once you go through the correct reference implementation, you'll see that they've used functions such as `GET_FLOAT_WORD` and `SET_FLOAT_WORD`, for single-precision numbers. We already have functions such as [`from-word`](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/number/float32/base/from-word) and [`to-word`](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/number/float32/base/to-word), which can be used here.

Neerajpathak07 replied:
Alright I'll go through the documentation of these first and have a good understanding about it.

Neerajpathak07 commented on lib/node_modules/@stdlib/math/base/special/powf/test/test.native.js:
have kept this empty as of now thinking of generating fixtures first and get back to this.

Neerajpathak07 commented on lib/node_modules/@stdlib/math/base/special/powf/README.md:
````diff
@@ -0,0 +1,240 @@
+
+```javascript
+var randu = require( '@stdlib/random/base/randu' );
+var round = require( '@stdlib/math/base/special/round' );
````
```suggestion
var roundf = require( '@stdlib/math/base/special/roundf' );
```
checked the dependency it should be `roundf` through will make the change soon.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/powf/lib/main.js:
```diff
@@ -0,0 +1,384 @@
+*
+* ## Notice
+*
+* The following copyright and license were part of the original implementation available as part of [FreeBSD]{@link https://svnweb.freebsd.org/base/release/9.3.0/lib/msun/src/s_pow.c}. The implementation follows the original, but has been modified for JavaScript.
```
@Neerajpathak07 Are you trying to port the double-precision implementation of `freebsd` to single-precision? That might not be the right way. You must be following the single-precision implementation directly. See: https://github.com/freebsd/freebsd-src/blob/main/lib/msun/src/e_powf.c

Neerajpathak07 replied:
I was referring to the upstream implementation of `powf` only but forgot to update the link here. I'll go through the implementation once again and start filling the gaps.

Neerajpathak07 commented:
## Updates so far cc: @kgryte @gunjjoshi 

- Went through the binary format for `float32` as discussed in the meet and referred documentation for `to-uint32`, `from-binary-string` & `to-binary-string` also studied the documentation for `constants/float32` packages.
- Changed var names across `logx` `logax` and `pow2` with accurate hexadecimal values based on upstream implementation with corresponding logical var names.
- Based all the JS implementation wrt to [powf](https://svnweb.freebsd.org/base/release/12.2.0/lib/msun/src/e_powf.c?revision=367086&view=markup) 
- used JS var like `towordf` and `forWordf` to resonate with `GET_FLOAT_WORD` & `SET_FLOAT_WORD`.
- reduced the `LOG_WORKSPACE` from an array of high and low bits to a singular var to compensate 32-bit int!!
- wrapped the implementation in `float64ToFloat32`
- reduced CI errors now most of them require C implementation of `powf`
- updated reference link in JS licenses to the `msun` file for `powf`
- removed the logic for high and low words space and added logic for single variable workspace in `main.js`

## Upcoming updates

- [ ] adding test fixtures.
- [ ] Working on test cases.
- [ ] adding C implementations once the JS one is approved.
- [ ] Fixing bugs and typo errors.

Neerajpathak07 commented on lib/node_modules/@stdlib/math/base/special/powf/lib/pow2.js:
```diff
@@ -34,26 +34,32 @@
 
 // VARIABLES //
 
 // 0x00100000 = 1048576 => 0 00000000001 00000000000000000000 => biased exponent: 1 = -1022+1023 => 2^-1022
```
will change the hexadecimal representation once the variables are approved.
Same for `HIGH_BIASED_EXP_NEG_1`

Neerajpathak07 commented on lib/node_modules/@stdlib/math/base/special/powf/lib/log2ax.js:
```diff
@@ -0,0 +1,226 @@
+
+// VARIABLES //
+
+// 0x007fffff = 8388607 => 0 00000000000 11111111111111111111
+var HIGH_EXP_2 = 0x007fffff; // asm type annotation
```
Will be using `FLOAT32_SIGNIFICAND_MASK` instead here.

Neerajpathak07 commented on lib/node_modules/@stdlib/math/base/special/powf/lib/pow2.js:
```diff
@@ -0,0 +1,146 @@
+// 0xffff8000 = 4294934528 => 1 11111111111 11111000000000000000
+var HIGH_NEG_NAN = 0xffff8000; // asm type annotation
+
+var HIGH_NUM_SIGNIFICAND_BITS = 23; // asm type annotation
```
Thought of creating a var manually to return 23 as `var bits` since couldn't find any binary component for this.

gunjjoshi replied:
@Neerajpathak07 You can use https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/constants/float32/precision here. Maybe you can use it as `FLOAT32_PRECISION - 1`, as the number of significand bits is 23. The value of `FLOAT32_PRECISION` is 24, as 1 hidden bit is included there too.

Neerajpathak07 commented on lib/node_modules/@stdlib/math/base/special/powf/lib/main.js:
```diff
@@ -0,0 +1,358 @@
+var float64ToFloat32 = require( '@stdlib/number/float64/base/to-float32' );
+var NINF = require( '@stdlib/constants/float32/ninf' );
+var PINF = require( '@stdlib/constants/float32/pinf' );
+var ABS_MASK = require( '@stdlib/constants/float64/high-word-abs-mask' );
```
forgot to update this 
```suggestion
var FLOAT32_ABS_MASK = require( '@stdlib/constants/float32/abs-mask' );
```

Neerajpathak07 replied:
to return `0x7fffffff`

Neerajpathak07 commented on lib/node_modules/@stdlib/math/base/special/powf/lib/main.js:
```diff
@@ -0,0 +1,358 @@
+		// Over- or underflow if `x` is not close to unity...
+		if ( ahx < HIGH_MAX_NEAR_UNITY ) {
+			// y < 0
+			if ( sn === 1 ) {
+				// Signal overflow...
+				return sn * HUGE * HUGE;
+			}
+			// Signal underflow...
+			return sn * TINY * TINY;
```
@gunjjoshi was a little confused about handling Over and underflow when `x` is not close to unity

Neerajpathak07 committed:
chore: update modules & var

Neerajpathak07 committed:
chore: update modules & var

Neerajpathak07 commented:
Had to rebase a branch as an unwanted change was added Ik it's unethical to do this but it was messing the purpose of this big-PR everything is running clean now working on more features.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/powf/lib/main.js:
```diff
@@ -0,0 +1,354 @@
+*
+* ## Notice
+*
+* The following copyright and license were part of the original implementation available as part of [FreeBSD]{@link https://svnweb.freebsd.org/base/release/9.3.0/lib/msun/src/e_powf.c}. The implementation follows the original, but has been modified for JavaScript.
```
We can update this `9.3.0` version to use the latest implementation.

gunjjoshi replied:
We can use this: https://svnweb.freebsd.org/base/release/12.2.0/lib/msun/src/e_powf.c

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/powf/lib/main.js:
```diff
@@ -0,0 +1,354 @@
+/**
+* Evaluates the exponential function for single-precision values.
+*
+* ## Method
```
`Method` is not required here. We can remove this, as it is not present in the reference implementation.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/powf/lib/main.js:
````diff
@@ -0,0 +1,354 @@
+*     x^y = 2^n e^{y^\prime \cdot \mathrm{log2}}
+*     ```
+*
+* ## Special Cases
````
Similarly, this and the `Notes` section below can be removed too.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/powf/lib/main.js:
```diff
@@ -0,0 +1,354 @@
+	var z; // y prime
+	var j;
+
+	hx = fromWordf( x );
+	hy = fromWordf( y );
```
@Neerajpathak07 `hx` and `hy` are words, which we are getting from the numbers `x` and `y`. So, we should be using `toWordf` here, instead of `fromWordf`.

Neerajpathak07 replied:
`msun` implementation stated 
```C
GET_FLOAT_WORD(hx,x);
GET_FLOAT_WORD(hy,y);
```

Neerajpathak07 replied:
which is why thought of using `fromWordf`

gunjjoshi commented:
@Neerajpathak07 Let's follow https://svnweb.freebsd.org/base/release/12.2.0/lib/msun/src/e_powf.c?revision=367086&view=markup here. The current implementation in `main.js` looks a bit off from this reference. You might want to restructure it, so that it matches with the reference.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/powf/lib/main.js:
```diff
@@ -0,0 +1,354 @@
+	}
+
+	// Special cases `x`...
+	if ( lx === 0 ) {
```
Where is this `lx` coming from?

gunjjoshi replied:
Seems like you've done a port of the double-precision implementation. Ideally, we must be independently following the single-precision reference, as the implementations may vary significantly.

Neerajpathak07 replied:
@gunjjoshi I haven't started refactoring the `main.js` yet was working on `log2ax`, `logx` and `pow2` to get them in line with single-precision implementation and wanted to get approvals on the var names and hexadecimal values first.
What I have changed in `main.js` is:-

- Removed the logic for high and low words.
- Align the var with correct package convention.

gunjjoshi replied:
@Neerajpathak07 I would have started the implementation from `main.js`, referring to the freeBSD implementation, rather than first working on `log2ax`, `logx` and `pow2`.

Neerajpathak07 replied:
Noted I'll keep that in mind.

Neerajpathak07 committed:
chore: refactoring js implementation

Neerajpathak07 committed:
chore: removing lx and hx from special cases

anandkaranubc commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into add-powf

anandkaranubc committed:
docs: update function description

anandkaranubc committed:
fix: correct the JS implementation of `powf`

anandkaranubc committed:
docs: fix spacing

anandkaranubc committed:
feat: add C implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
chore: remove extra variable

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
chore: reorder if statements

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
docs: clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
chore: update copyright years

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
chore: reorder if statements

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
chore: fix package name

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc reviewed (dismissed):
Thanks @Neerajpathak07 for your contribution! This will indeed be a great addition! 🚀

Neerajpathak07 commented:
@anandkaranubc Thanks for your continuous efforts to get this one over the finish line.

Planeshifter reviewed (dismissed):
LGTM! Thank you @Neerajpathak07 and @anandkaranubc!

There are various discrepancies between the function parameter names in the implementation and public-facing docs, but I do no think we need to address these. They are also present in `pow`.

Neerajpathak07 commented:
@kgryte could we land this in?

kgryte commented on lib/node_modules/@stdlib/math/base/special/powf/lib/logxf.js:
```diff
@@ -0,0 +1,100 @@
+	t1 = f32( u + v );
+	tmp = toWordf( t1 ) | 0; // asm type annotation
+	t1 = fromWordf( tmp & TRUNC_MASK );
+	t2 = f32( f32( v ) - f32( t1 - u ) );
```
```suggestion
	t2 = f32( v - f32( t1 - u ) );
```

kgryte committed:
refactor: remove unnecessary wrapping

`v` is already float32 per L86.

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/math/base/special/powf/package.json:
```diff
@@ -0,0 +1,190 @@
+          "rand": {
+            "prng": "random/base/uniform",
+            "parameters": [
+              -10,
```
```suggestion
              0,
```

kgryte committed:
chore: keep exponent non-negative to prevent NaNs

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM
